cni/network/invoker_cns.go

```diff
@@ -70,94 +72,143 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
-		return IPAMAddResult{}, errors.Wrap(err, "Failed to get IP address from CNS with error: %w")
+		// checks for 404 error
+		if errors.Is(err, cnscli.ErrAPINotFound) {
+			ipconfigRequest := cns.IPConfigRequest{
+				OrchestratorContext: orchestratorContext,
+				PodInterfaceID:      GetEndpointID(addConfig.args),
+				InfraContainerID:    addConfig.args.ContainerID,
+			}
+			log.Errorf("Failed to request IPs using RequestIPs from CNS, going to try RequestIPAddress. error: %v request: %v", err, ipconfig)
+
+			res, err := invoker.cnsClient.RequestIPAddress(context.TODO(), ipconfigRequest)
+			if err != nil {
+				// if the old API fails as well then we just return the error
+				log.Errorf("Failed to release IP address from CNS using ReleaseIPAddress. error: %v request: %v", err, ipconfigRequest)
+				return IPAMAddResult{}, errors.Wrap(err, fmt.Sprintf("failed to request IPs for pod %v with error ", GetEndpointID(addConfig.args))+"%w")
+			}
+			// sets the response to the new contract
+			response = &cns.IPConfigsResponse{
+				Response: res.Response,
+				PodIPInfo: []cns.PodIpInfo{
+					res.PodIpInfo,
+				},
+			}
+
+		} else {
+			log.Printf("Failed to get IP address from CNS with error %v, response: %v", err, response)
+			return IPAMAddResult{}, errors.Wrap(err, "Failed to get IP address from CNS with error: %w")
+		}
 	}
 
-	info := IPv4ResultInfo{
-		podIPAddress:       response.PodIpInfo.PodIPConfig.IPAddress,
-		ncSubnetPrefix:     response.PodIpInfo.NetworkContainerPrimaryIPConfig.IPSubnet.PrefixLength,
-		ncPrimaryIP:        response.PodIpInfo.NetworkContainerPrimaryIPConfig.IPSubnet.IPAddress,
-		ncGatewayIPAddress: response.PodIpInfo.NetworkContainerPrimaryIPConfig.GatewayIPAddress,
-		hostSubnet:         response.PodIpInfo.HostPrimaryIPInfo.Subnet,
-		hostPrimaryIP:      response.PodIpInfo.HostPrimaryIPInfo.PrimaryIP,
-		hostGateway:        response.PodIpInfo.HostPrimaryIPInfo.Gateway,
-	}
+	addResult := IPAMAddResult{}
 
-	// set the NC Primary IP in options
-	addConfig.options[network.SNATIPKey] = info.ncPrimaryIP
+	for i := 0; i < len(response.PodIPInfo); i++ {
+		info := IPResultInfo{
+			podIPAddress:       response.PodIPInfo[i].PodIPConfig.IPAddress,
+			ncSubnetPrefix:     response.PodIPInfo[i].NetworkContainerPrimaryIPConfig.IPSubnet.PrefixLength,
+			ncPrimaryIP:        response.PodIPInfo[i].NetworkContainerPrimaryIPConfig.IPSubnet.IPAddress,
+			ncGatewayIPAddress: response.PodIPInfo[i].NetworkContainerPrimaryIPConfig.GatewayIPAddress,
+			hostSubnet:         response.PodIPInfo[i].HostPrimaryIPInfo.Subnet,
+			hostPrimaryIP:      response.PodIPInfo[i].HostPrimaryIPInfo.PrimaryIP,
+			hostGateway:        response.PodIPInfo[i].HostPrimaryIPInfo.Gateway,
+		}
 
-	log.Printf("[cni-invoker-cns] Received info %+v for pod %v", info, podInfo)
+		// set the NC Primary IP in options
+		addConfig.options[network.SNATIPKey] = info.ncPrimaryIP
```

Comment:
this value will get overwritten with the for loop. Where is this used? what happens if it gets ipv6 ip?

Reply:
Hi, I am not sure about this since ncPrimaryIP is from PodIPInfo IPAddress, can we discuss this later on